File: tests/bridges/lua_scripts/t_config_text.lua
local atp = require("atp")

local M = atp.module("lua_config_text", { 1, 0 })

M.out_report = atp.output(atp.text)

function M:initialize()
    local rate = "none"
    if self.config ~= nil and self.config.audio ~= nil and self.config.audio.rate ~= nil then
        rate = tostring(self.config.audio.rate)
    end
    local origin = "none"
    if self.config_origin ~= nil and self.config_origin ~= "" then
        origin = self.config_origin
    end
    self.report = string.format("text-len=%d origin=%s opaque=%s rate=%s", #self.config_text, origin,
        tostring(self.config_opaque), rate)
end

function M:iterate()
    if self.report == nil then
        return atp.IDLE
    end
    self.out_report:write(self.report)
    self.report = nil
    return atp.BUSY
end

return M

File: tests/bridges/scripts/t_config_text.py
"""Reads the raw text of a config the host did not parse, which is the point of an opaque format.

The three attributes are read in the constructor for the same reason self.config is: a module decides
what it is before initialize, and a format the platform does not know is decided here or nowhere.
"""

import atp


class ConfiguredByText(atp.Module):
    name = "py_config_text"

    out_report = atp.Output(atp.text)

    def __init__(self):
        self.report = "text-len={} origin={} opaque={} rate={}".format(
            len(self.config_text),
            self.config_origin or "none",
            self.config_opaque,
            (self.config or {}).get("audio", {}).get("rate", "none"),
        )

    def iterate(self):
        if self.report is None:
            return atp.IDLE
        self.out_report.write(self.report)
        self.report = None
        return atp.BUSY
